engine/proxy.go

```diff
@@ -0,0 +1,227 @@
+package engine
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/prisma/prisma-client-go/binaries"
+	"github.com/prisma/prisma-client-go/logger"
+	"github.com/prisma/prisma-client-go/runtime/types"
+)
+
+func NewDataProxyEngine(schema, schemaEnvVarName string) *DataProxyEngine {
+	return &DataProxyEngine{
+		Schema:           schema,
+		schemaEnvVarName: schemaEnvVarName,
+		http:             &http.Client{},
+	}
+}
+
+type DataProxyEngine struct {
+	// http is the internal http client
+	http *http.Client
+
+	// url holds the query-engine url
+	url string
+
+	// schemaEnvVarName is the env var for the datasource url
+	// this is needed internally to extract the api key
+	schemaEnvVarName string
+
+	// Schema contains the prisma Schema
+	Schema string
+
+	// apiKey contains the parsed prisma data proxy api key from the connection string
+	apiKey string
+}
+
+func (e *DataProxyEngine) Connect() error {
+	// Example uri: https://aws-eu-west-1.prisma-data.com/2.26.0/412bf0a1742a576d699fbd5102a4f725557eff3992995f2e18febce128794961/
+	hash := hashSchema(e.Schema)
+	logger.Debug.Printf("local schema hash %s", hash)
+
+	logger.Debug.Printf("parsing connectino string from database url %s", e.schemaEnvVarName)
+
+	connectionString := os.Getenv(e.schemaEnvVarName)
+	if connectionString == "" {
+		return fmt.Errorf("no connection string found")
+	}
+	u, err := url.Parse(connectionString)
+	if err != nil {
+		return fmt.Errorf("parse prisma string: %w", err)
+	}
+	e.apiKey = u.Query().Get("api_key")
+	if e.apiKey == "" {
+		return fmt.Errorf("could not parse api key from data proxy prisma connection string")
+	}
+
+	e.url = getCloudURI(u.Host, hash)
+	// TODO temp: remove this once prisma data proxy works with recent stable versions – right now only 3.0.1 works
+	e.url = "https://" + path.Join(u.Host, "3.0.1", hash)
+	logger.Debug.Printf("using %s as remote URI", e.url)
+	if err := e.uploadSchema(context.Background()); err != nil {
```

Comment:
From what I remember TS client uploads the schema when it needs to - aka, it tries to perform  the request first. 
I think this diff is correct for the use case of Go, where the server is likely to be longer lived, and unlikely to be a lambda.
Perhaps it'd be worth a comment somewhere noting the difference between the two clients.